### ISSUE TITLE
Interested in resolving issues from the SMILES benchmark?

### ISSUE THREAD

**Issue by baoilleach** [epam/Indigo#129]
I've put together a benchmark to compare the ability of different tools to read SMILES. The goal is to identify corner cases and improve SMILES reading across the board. Initial results were reported at the recent ICCS:
https://nextmovesoftware.com/blog/2018/06/06/can-we-agree-on-the-structure-represented-by-a-smiles-string/
...and the benchmark itself is available at:
https://github.com/nextmovesoftware/smilesreading

In the benchmark, each toolkit reports the number of hydrogens on each atom in the molecule represented by an aromatic SMILES string, or whether there was an error reading it. If you are interested in working through some of the issues raised by the benchmark, let me know and I am happy to describe them and how they might be solved. I have already done this with a number of toolkits.

**Comment by AlexanderSavelyev** [comment]
yes, it would be great to know the issues for the SMILES test. Could you share details

**Comment by baoilleach** [comment]
Excellent. I'll describe some of the issues found when Indigo is used to read aromatic SMILES strings generated by the CDK. This corresponds to "Comparison 1" in the ICCS poster (shown in the blog post above). I tested Indigo 1.3.0b.r16.

These are in the format of:
```
SMILES benchmark_id
benchmark_id [implicit hydrogen counts for each atom in order] [Program tested]
```

```
n1cc-2c[nH]n(c2n1)C 3202
3202 0 1 0 1 1 0 0 0 3 CDK
3202 1 2 1 2 1 0 1 1 3 Indigo

c1ccc2-c(nnc2)o1 4181
4181 1 1 1 0 0 0 0 1 0 CDK
4181 2 2 2 1 1 1 1 2 0 Indigo

n1c2-c(cnn2)cc[nH]1 4201
4201 0 0 0 1 0 0 1 1 1 CDK
4201 1 1 1 2 1 1 2 2 1 Indigo

c-12ccc[nH]c1nnn2 4509
4509 0 1 1 1 1 0 0 0 0 CDK
4509 1 2 2 2 1 1 1 1 1 Indigo
```

I've had a quick look at the approach Indigo uses to read aromatic SMILES, and it appears similar to that used by the current release of Open Babel, to attempt to find aromatic systems and dearomatise. I've since replaced this with the approach used by Daylight, which is to attempt to find a "perfect matching". You might want to consider this at some point.

**Comment by IuriiPuzanov** [comment]
Hello, 

  I've checked these examples with current Indigo version (master branch, Indigo 1.3.0beta.r364.) and it produces the same results as CDK. The only requirement is using new option "ignore-bad-valence" to avoid issues with loading such structures.
  It would be nice to check all other wrong cases from your benchmark. 

Best Regards!
Iurii

**Comment by baoilleach** [comment]
I'm a bit confused - there is no "bad valence" here. Here is an image of Kekule forms of these structures as generated by http://www.simolecule.com/cdkdepict/depict.html:
![capture](https://user-images.githubusercontent.com/45337/42763402-d91e2b0a-890a-11e8-8cde-18a697b07458.PNG)

**Comment by IuriiPuzanov** [comment]
Yes, you are right and actually this new flag is required only for calculation of implicit hydrogens in the aromatic form of these molecules. If calculation of implicit hydrogens are made for dearomatized from then current Indigo version produces the same results as CDK without any additional flags.

Best Regards!
Iurii

**Comment by IuriiPuzanov** [comment]
No, at the moment this flag is required only for calculation implicit hydrogens (which is used for comparison) in some special cases (like your examples). It is not required for reading aromatic SMILES.

Best Regards!
Iurii

**Comment by baoilleach** [comment]
So I should always specify this option if I wish to have the correct number of implicit hydrogens in the molecule read from an aromatic SMILES string?

**Comment by baoilleach** [comment]
I guess the reason I keep coming back to this issue is that counting the hydrogens on one of these strings is fairly straightforward, and so Indigo should be able to always get this correct without the user needing to specify any option.

In general you round up to 3 or 5 for nitrogen and 4 for carbon. For aromatic atoms, it's easiest to round up to 2 or 4 for nitrogen and 3 for carbon and consider aromatic bonds as single bonds.

With this in mind, I can calculate by hand the hydrogen counts for each atom in the SMILES string "n1cc-2c[nH]n(c2n1)C 3202":
```
1. n - 2 aromatic bonds, round up to 2, gives 0
2. c - 2 aromatic bonds, round up to 3, gives 1
3. c - 2 aromatic bonds + 1 single bond, round up to 3, gives 0
4. c - 2 aromatic bonds, round up to 3, gives 1
5. [nH] - explicitly given as 1
6. n - 2 aromatic bonds + 1 single bond, round up to 2, gives 0
7. c - 2 aromatic bonds + 1 single bond, round up to 3, gives 0
8. n - 2 aromatic bond, round up to 2, gives 0
9. C - 1 single bond, round up to 4, gives 3
```
